### PR TITLE
fixes |mass

### PR DESCRIPTION
**Pull request by joemfb**
refactors and unifies memory marking implementation for GC and `|mass`

### CODE REVIEW

File: noun/allocate.c [hunk 1]
```diff
@@ -1628,22 +1628,128 @@ u3a_print_memory(c3_c* cap_c, c3_w wor_w)
...
+  if ( u3a_minimum < box_u->siz_w ) {
+    // fprintf(stderr, "print box 1\r\n");
+    return u3m_pretty(u3a_to_pug(u3a_outa(u3a_boxto(box_u))));
+  }
+  else {
+    // this might not be a cell, but we can use the struct to inspect further
+    //
+    u3a_cell* cel_u = (u3a_cell*)box_u;
+
+    // this condition potentially corresponds to an indirect atom of only one word
```
Comment (philipcmonk) [inline noun/allocate.c]: I think it's worth explicitly noting somewhere that if this condition is false, we're sure this is a cell.

File: noun/allocate.c [hunk 2]
```diff
@@ -1628,22 +1628,128 @@ u3a_print_memory(c3_c* cap_c, c3_w wor_w)
...
+    //
+    if ( (1 == (c3_w)cel_u->hed) &&
+         (0x80000000 & (c3_w)cel_u->tel) )
+    {
+      // this might not be right, but we know it's safe
```
Comment (philipcmonk) [inline noun/allocate.c]: not clear what is "this".  Perhaps:  "this might not be an indirect atom, but we know it's safe to print as one".

**Comment by joemfb** [comment]
I'm testing these printfs further, making sure they work with arbitrary structures and large amounts of data. I'll push some fixes soon.

**Comment by ixv** [comment]
Can we save this output to a file? It would work nicer with this: https://github.com/urbit/arvo/pull/1028

Maybe we should #define LOGGING  or something for things like this.

**Comment by belisarius222** [comment]
I feel like ideally we'd be uploading this data to https://docs.datadoghq.com/graphing or something similar.

**Comment by joemfb** [comment]
I cleaned up the comments, and disabled printing of atoms larger than 10 words. The allocation-box heuristic printing is still crummy, but it's quite a bit better than it was.

@ixv, I could thread a `File*` pointer instead of a loobean through all the mark functions -- that could default to `stderr` but pointed to some external file through a build option or something.

**Comment by joemfb** [comment]
@ixv, configure meson with `-Dmemory-log=true` and every `|mass` effect will be written to `pier/.urb/put/$DATE-mass.txt`.

**Comment by joemfb** [comment]
@philipcmonk, please re-review.

**Pull request merged**